### ISSUE TITLE
pendingCount < 0 means testability will never be ready

### ISSUE THREAD

**Issue by henrahmagix** [alfonso-presa/testability.js#4]
If `pendingCount` becomes less than zero, because `wait.oneLess()` is called too many times, then the `pendingCallbacks` will never be called.

**Comment by alfonso-presa** [comment]
While this issue is completely true, I wonder what is the situation that can lead to this to happen. You're supposed to call `wait.oneLess()` only once when a task that has been started with `wait.oneMore()` ends. Do you have scenario? Is this happening because testability.js is not present when `testability && testability.wait.oneMore()` is being invoked and it is when `testability && testability.wait.oneLess()` is?

Thanks!

**Comment by alfonso-presa** [comment]
Thinking of this, it might not be a good idea to expose `oneMore` and `oneLess` as public API. I'm think that a better API for people not wanting to use the `wait.for` mechanism might be something like:

``` js
var task = testability && testability.wait.start();

/* async things happening */

task && task.end();
```

This way you can only end tasks you have started leading to negative count to be impossible.

What do you think?

**Comment by henrahmagix** [comment]
@alfonso-presa I like your `wait.start()` prototype, I think that would be a nice addition. In my specific case though, I don't think that would work. 

> You're supposed to call `wait.oneLess()` only once when a task that has been started with `wait.oneMore()` ends.

In my ideal scenario, it would be exactly as you describe =D And in all my other projects it is just like that. Let me say that I really like this plugin! It's a fantastic and simple thing that's incredibly helpful for using Protractor on non-Angular pages.

> Is this happening because testability.js is not present when `testability && testability.wait.oneMore()` is being invoked and it is when `testability && testability.wait.oneLess()` is?

Fortunately that's not the case: testability is always there, it's just a matter of when I'm using it.

> Do you have scenario?

**tl;dr: I can't store variables across the contexts I am testing.**

I am using the [Cobalt Engine](https://cobalt-engine.com/documentation) slide presentation framework, which bundles together scripts for each slide and each slide is in an `<iframe>` so they don't share the same global scope. Therefore I need to access `window.parent.testability` inside the slide scripts, and slides can't easily talk to each other.

I am listening for the `willgoto` (triggers when a swipe is started) and `slideenter` (triggers when a slide is focused) slide events, so I can `wait.oneMore()` in `willgoto` and `wait.oneLess()` in `slideenter` to allow Protractor to swipe a slide and wait for the next slide to come in before continuing with the control flow.

However, sometimes `slideenter` is triggered when the presentation loads, and I have no way of persisting variables between slides, so I can't stop `wait.oneLess()` from triggering more often than it should. I could possibly use [`data-exchange`](https://cobalt-engine.com/documentation/index#5629fc530fd3530037576648) but that only initialises in the `nav.onenter()` event, which I think is too late for `willgoto`.

**Comment by henrahmagix** [comment]
I've tried persisting a boolean between slides (via `window.parent.waitingForSlide = true`), but the timing of the `willgoto` and `slideenter` events don't quite work with this.
